best-practices.md

```diff
@@ -0,0 +1,183 @@
+            }
+        }
+    }
+
+    @Override
+    public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+        phaser = new Phaser(N + 1); // +1 for self
+        phaser.arriveAndAwaitAdvance();
+        /* close file being written */
+    }
+
+    @Override
+    public void afterRestore(Context<? extends Resource> context) throws Exception {
+        Phaser phaser = this.phaser;
+        this.phaser = null;
+        try {
+            /* reopen the file */
+            phaser.arriveAndAwaitAdvance();
+        } catch (Exception e) {
+            phaser.forceTermination();
+            throw e;
+        }
+    }
+}
+```
+
+This synchronization requires only one volatile read on each `write()` call, that is generally a cheap operation. However if one of the expected threads is waiting for a long time the checkpoint would be blocked. This could be mitigated by using shorter timeouts (e.g. if the thread is polling a queue) or even actively interrupting it from the `beforeCheckpoint` method.
+
+## Eventloop model
+
+The best case is the eventloop model where the application uses single thread for all operations in the resource and already has a mechanism to schedule task in that eventloop. Let's take an example of a resource sending a heartbeat message.
+
+```java
+public class HeartbeatManager implements Runnable, Resource {
+    public final ScheduledExecutorService eventloop; // single-threaded
+    public boolean suspended;
+
+    public HeartbeatManager(Executor eventloop) {
+        eventloop.scheduleAtFixedRate(this, 0, 1, TimeUnit.MINUTES);
+    }
+
+    @Override
+    public void run() {
+        /* send heartbeat message */
+    }
+
+    @Override
+    public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+        synchronized (this) {
+            HeartbeatManager self = this;
+            executor.execute(() -> {
+                synchronized (self) {
+                    self.suspended = true;
+                    self.notify();
+                    while (self.suspended) {
+                        self.wait();
+                    }
+                }
+            })
+            while (!suspended) {
+                wait();
+            }
+        }
+        /* shutdown */
+    }
+
+    @Override
+    public void afterRestore(Context<? extends Resource> context) throws Exception {
+        /* restore */
+        synchronized (this) {
+            suspended = false;
+            notify();
+        }
+    }
+}
+```
+
+Note one detail in the example above: if the application is stopped for a long time the task scheduled by the `ScheduledExecutorService.scheduleAtFixedRate(...)`  would try to keep up after restore and perform all the missed invocations. Handling that should be a part of the `beforeCheckpoint` procedure, cancelling the task and rescheduling it again in `afterRestore`.
+
+## Implementing Resource as inner class
```

Comment:
It would be nice to structure this section out from the list of the example patterns.

Does it make sense to move them to subsections?

* Implementing Resource as inner class
* Common CRaC Patterns
  * Unknown number of threads arriving randomly
  * One or known number of periodically arriving threads
  * ...

?